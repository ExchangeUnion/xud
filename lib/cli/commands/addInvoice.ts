import { callback, loadXudClient } from '../command';
import { Arguments } from 'yargs';
import { AddInvoiceRequest } from '../../proto/xudrpc_pb';

export const command = 'addinvoice <value> [memo]';

export const describe = 'get orders from the order book';

export const builder = {
  value: {
    type: 'number',
  },
  memo: {
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new AddInvoiceRequest();
  request.setValue(argv.value);
  request.setMemo(argv.memo);
  loadXudClient(argv).addInvoice(request, callback);
};
