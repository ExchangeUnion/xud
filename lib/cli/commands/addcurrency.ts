import { Arguments } from 'yargs';
import { callback, loadXudClient } from '../command';
import { AddCurrencyRequest } from '../../proto/xudrpc_pb';
import { SwapClients } from '../../types/enums';

export const command = 'addcurrency <currency> <swap_client> [decimal_places] [token_address]';

export const describe = 'add a currency';

export const builder = {
  currency: {
    description: 'The ticker symbol for the currency',
    type: 'string',
  },
  swap_client: {
    description: 'The payment channel network client for swaps',
    type: 'string',
    choices: ['LND', 'RAIDEN'],
  },
  decimal_places: {
    description: 'The places to the right of the decimal point of the smallest subunit (e.g. satoshi)',
    default: 8,
    type: 'number',
  },
  token_address: {
    description: 'The contract address for layered tokens such as ERC20',
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new AddCurrencyRequest();
  request.setCurrency(argv.currency.toUpperCase());
  request.setSwapClient(Number(SwapClients[argv.swap_client]));
  request.setTokenAddress(argv.token_address);
  request.setDecimalPlaces(argv.decimal_places);
  loadXudClient(argv).addCurrency(request, callback);
};
