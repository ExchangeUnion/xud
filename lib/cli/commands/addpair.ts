import { Arguments } from 'yargs';
import { callback, loadXudClient } from '../command';
import { AddPairRequest } from '../../proto/xudrpc_pb';

export const command = 'addpair <base_currency> <quote_currency>';

export const describe = 'add a trading pair';

export const builder = {
  base_currency: {
    description: 'The currency bought and sold for this trading pair',
    type: 'string',
  },
  quote_currency: {
    description: 'The currency used to quote a price',
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new AddPairRequest();
  request.setBaseCurrency(argv.base_currency.toUpperCase());
  request.setQuoteCurrency(argv.quote_currency.toUpperCase());
  loadXudClient(argv).addPair(request, callback);
};
