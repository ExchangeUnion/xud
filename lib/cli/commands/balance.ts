import { callback, loadXudClient } from '../command';
import { Arguments } from 'yargs';
import { BalanceRequest, BalanceResponse } from '../../proto/xudrpc_pb';
import { satsToCoinsStr } from '../utils';
import Table, { HorizontalTable } from 'cli-table3';
import colors from 'colors/safe';

const HEADERS = [
  colors.blue('Ticker'),
  colors.blue('Balance'),
  colors.blue('Pending Balance'),
];

const formatBalances = (balances: BalanceResponse.AsObject) => {
  const formatted: any[] = [];
  balances.balancesMap.forEach((balance) => {
    const element = [];
    element.push(balance[0], `${satsToCoinsStr(balance[1].balance)}`, `${satsToCoinsStr(balance[1].pendingOpenBalance)}`);
    formatted.push(element);
  });
  return formatted;
};

const createTable = () => {
  const table = new Table({
    head: HEADERS,
  }) as HorizontalTable;
  return table;
};

const displayBalances = (balances: BalanceResponse.AsObject) => {
  const table = createTable();
  const formatted = formatBalances(balances);
  formatted.forEach(balance => table.push(balance));
  console.log(colors.underline(colors.bold('\nBalance:')));
  console.log(table.toString());
};

export const command = 'balance [currency]';

export const describe = 'get total balance for a given currency';

export const builder = {
  currency: {
    describe: 'the currency to query for',
    type: 'string',
  },
};

export const handler = (argv: Arguments) => {
  const request = new BalanceRequest();
  if (argv.currency) {
    request.setCurrency(argv.currency.toUpperCase());
  }
  loadXudClient(argv).balance(request, callback(argv, displayBalances));
};
