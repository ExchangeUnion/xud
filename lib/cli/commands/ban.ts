import { Arguments, Argv } from 'yargs';
import { BanRequest } from '../../proto/xudrpc_pb';
import { callback, loadXudClient } from '../command';

export const command = 'ban <node_identifier>';

export const describe = 'ban a remote node';

export const builder = (argv: Argv) =>
  argv
    .positional('node_identifier', {
      description: 'the node key or alias of the remote node to ban',
      type: 'string',
    })
    .example('$0 ban 028599d05b18c0c3f8028915a17d603416f7276c822b6b2d20e71a3502bd0f9e0b', 'ban by node key')
    .example('$0 ban CheeseMonkey', 'ban by alias');

export const handler = async (argv: Arguments<any>) => {
  const request = new BanRequest();
  request.setNodeIdentifier(argv.node_identifier);
  (await loadXudClient(argv)).ban(request, callback(argv));
};
