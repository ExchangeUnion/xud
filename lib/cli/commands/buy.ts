import { Arguments, Argv } from 'yargs';
import { OrderSide } from '../../proto/xudrpc_pb';
import { placeOrderBuilder, placeOrderHandler } from '../placeorder';

export const command = 'buy <quantity> <pair_id> <price> [order_id]';

export const describe = 'place a buy order';

export const builder = (argv: Argv) => placeOrderBuilder(argv, OrderSide.BUY);

export const handler = async (argv: Arguments) => placeOrderHandler(argv, OrderSide.BUY);
