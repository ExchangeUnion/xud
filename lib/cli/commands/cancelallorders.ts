import { Arguments, Argv } from 'yargs';
import { CancelAllOrdersRequest, CancelAllOrdersResponse } from '../../proto/xudrpc_pb';
import { callback, loadXudClient } from '../command';

export const command = 'cancelallorders';

export const describe = 'removes all orders';

export const builder = (argv: Argv) => argv
    .example('$0 cancelallorders', describe);

const formatOutput = (response: CancelAllOrdersResponse.AsObject) => {
  if (response.removedOrderIdsList.length) {
    response.removedOrderIdsList.forEach((removedOrder => console.log(`Cancelled order with id ${removedOrder}`)));
  } else {
    console.log('No orders found');
  }
};

export const handler = async (argv: Arguments<any>) => {
  (await loadXudClient(argv)).cancelAllOrders(new CancelAllOrdersRequest(), callback(argv, formatOutput));
};
