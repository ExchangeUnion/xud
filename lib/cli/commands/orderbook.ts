import Table, { HorizontalTable } from 'cli-table3';
import colors from 'colors/safe';
import { Arguments, Argv } from 'yargs';
import { OrderBookRequest, OrderBookResponse } from '../../proto/xudrpc_pb';
import { callback, loadXudClient } from '../command';
import { satsToCoinsStr } from '../utils';

const COLUMNS = [19, 19, 19, 19];
const HEADER = [
  { content: colors.green('Buy'), colSpan: 2 },
  { content: colors.red('Sell'), colSpan: 2 },
];
const SECONDARY_HEADER = [colors.green('Quantity'), colors.green('Price'), colors.red('Price'), colors.red('Quantity')];

let precision = 0;

const createTable = () => {
  const table = new Table({ colWidths: COLUMNS }) as HorizontalTable;
  table.push(HEADER);
  table.push(SECONDARY_HEADER);
  return table;
};

const displayOrderbook = (pairId: string, orderbook: OrderBookResponse.Buckets.AsObject) => {
  const table = createTable();
  const { buyBucketsList, sellBucketsList } = orderbook;
  const rowCount = Math.max(buyBucketsList.length, sellBucketsList.length);

  for (let n = 0; n < rowCount; n += 1) {
    const row: string[] = [];
    row[0] = buyBucketsList[n] ? satsToCoinsStr(buyBucketsList[n].quantity) : '';
    row[1] = buyBucketsList[n]?.price.toString() ?? '';
    if (row[1] === '0') {
      // instead of displaying zero, display the < symbol and the smallest price bucket
      // given the specified precision. for example if precision is 2, a zero value
      // would be converted to "<0.01"
      row[1] = `<${1 / 10 ** precision}`;
    }
    row[2] = sellBucketsList[n]?.price.toString() ?? '';
    row[3] = sellBucketsList[n] ? satsToCoinsStr(sellBucketsList[n].quantity) : '';
    table.push(row);
  }

  console.log(colors.underline(colors.bold(`\nTrading pair: ${pairId}`)));
  console.log(table.toString());
};

export const displayTables = (orderbooks: OrderBookResponse.AsObject) => {
  orderbooks.bucketsMap.forEach((val) => {
    displayOrderbook(val[0], val[1]);
  });
};

export const command = 'orderbook [pair_id] [precision]';

export const describe = 'display the order book, with orders aggregated per price point';

export const builder = (argv: Argv) =>
  argv
    .option('pair_id', {
      describe: 'trading pair for which to retrieve the order book',
      type: 'string',
    })
    .option('precision', {
      describe: 'the number of digits following the decimal point',
      type: 'number',
      default: 5,
    })
    .option('limit', {
      describe: 'the number of digits following the decimal point',
      type: 'number',
      default: 8,
    })
    .example('$0 orderbook', 'display the order books for all trading pairs')
    .example('$0 orderbook LTC/BTC', 'display the LTC/BTC order book')
    .example('$0 orderbook LTC/BTC 2', 'display the LTC/BTC order book with 2 decimal precision')
    .example('$0 orderbook --precision 2', 'display the order books for all trading pairs with 2 decimal precision')
    .example(
      '$0 orderbook BTC/USDT -2 --limit 10',
      'display the USDT/BTC order book with up to 10 buckets using hundreds place precision',
    );

export const handler = async (argv: Arguments<any>) => {
  const request = new OrderBookRequest();
  const pairId = argv.pair_id ? argv.pair_id.toUpperCase() : undefined;
  request.setPairId(pairId);
  request.setPrecision(argv.precision);
  precision = argv.precision; // store the precision to a global variable in case we need it for display purposes later
  request.setLimit(argv.limit);
  (await loadXudClient(argv)).orderBook(request, callback(argv, displayTables));
};
