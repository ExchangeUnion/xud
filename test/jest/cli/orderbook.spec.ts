import { displayTables } from '../../../lib/cli/commands/orderbook';

jest.mock('colors/safe', () => {
  return {
    green: (str: string) => str,
    red: (str: string) => str,
    bold: (str: string) => str,
    underline: (str: string) => str,
  };
});

const ltcBtcOrderBook = {
  buyBucketsList: [
    {
      price: 0.0099,
      quantity: 11071587640,
    },
    {
      price: 0.0098,
      quantity: 9117367602,
    },
    {
      price: 0.0097,
      quantity: 10614167904,
    },
    {
      price: 0.0096,
      quantity: 10266496213,
    },
    {
      price: 0.0095,
      quantity: 10398061003,
    },
    {
      price: 0.0094,
      quantity: 11980194419,
    },
    {
      price: 0.0093,
      quantity: 8204432528,
    },
    {
      price: 0.0092,
      quantity: 10357335925,
    },
    {
      price: 0.0091,
      quantity: 9471010321,
    },
    {
      price: 0.009,
      quantity: 8570250137,
    },
  ],
  sellBucketsList: [
    {
      price: 0.0101,
      quantity: 9797360538,
    },
    {
      price: 0.0102,
      quantity: 10006815933,
    },
    {
      price: 0.0103,
      quantity: 10292596794,
    },
    {
      price: 0.0104,
      quantity: 10770572898,
    },
    {
      price: 0.0105,
      quantity: 9506165727,
    },
    {
      price: 0.0106,
      quantity: 8786526193,
    },
    {
      price: 0.0107,
      quantity: 8987718204,
    },
    {
      price: 0.0108,
      quantity: 11081500717,
    },
    {
      price: 0.0109,
      quantity: 10279896939,
    },
    {
      price: 0.011,
      quantity: 9761906414,
    },
  ],
};

const orderbook: any = {
  bucketsMap: [['LTC/BTC', ltcBtcOrderBook]],
};

describe('orderbook', () => {
  const mockLog = jest.fn();
  console.log = mockLog;

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should print the orderbook', () => {
    displayTables(orderbook);
    expect(mockLog.mock.calls).toMatchSnapshot();
  });
});
