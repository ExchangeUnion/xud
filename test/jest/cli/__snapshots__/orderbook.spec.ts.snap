// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orderbook should print the orderbook 1`] = `
Array [
  Array [
    "
Trading pair: LTC/BTC",
  ],
  Array [
    "┌───────────────────────────────────────┬───────────────────────────────────────┐
│ Buy                                   │ Sell                                  │
├───────────────────┬───────────────────┼───────────────────┬───────────────────┤
│ Quantity          │ Price             │ Price             │ Quantity          │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 110.7158764       │ 0.0099            │ 0.0101            │ 97.97360538       │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 91.17367602       │ 0.0098            │ 0.0102            │ 100.06815933      │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 106.14167904      │ 0.0097            │ 0.0103            │ 102.92596794      │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 102.66496213      │ 0.0096            │ 0.0104            │ 107.70572898      │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 103.98061003      │ 0.0095            │ 0.0105            │ 95.06165727       │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 119.80194419      │ 0.0094            │ 0.0106            │ 87.86526193       │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 82.04432528       │ 0.0093            │ 0.0107            │ 89.87718204       │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 103.57335925      │ 0.0092            │ 0.0108            │ 110.81500717      │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 94.71010321       │ 0.0091            │ 0.0109            │ 102.79896939      │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 85.70250137       │ 0.009             │ 0.011             │ 97.61906414       │
└───────────────────┴───────────────────┴───────────────────┴───────────────────┘",
  ],
]
`;
